pkg/operator/v1helpers/helpers.go

```diff
@@ -347,3 +349,65 @@ func InjectObservedProxyIntoContainers(podSpec *corev1.PodSpec, containerNames [
 
 	return nil
 }
+
+func IsDeploymentProgressing(conditionName string, deployment *appsv1.Deployment, status *operatorv1.OperatorStatus) (bool, string) {
+	if deployment.Generation != deployment.Status.ObservedGeneration {
+		return true, "Waiting for Deployment to act on changes"
+	}
+
+	var expectedReplicas int32
+	if deployment.Spec.Replicas != nil {
+		expectedReplicas = *deployment.Spec.Replicas
+	}
+
+	if expectedReplicas > deployment.Status.UpdatedReplicas {
+		// Verify if the current deployment generation is the same as the one that this controller
+		// dealt with in the previous sync. If they are not the same, we're definitely rolling
+		// out new config, so we are progressing.
+		lastGeneration := resourcemerge.ExpectedDeploymentGeneration(deployment, status.Generations)
+		if deployment.Generation != lastGeneration {
+			return true, "Waiting for Deployment to deploy node pods"
+		}
+
+		// At this point we know two things: we already synced the current generation of the deployment,
+		// but the Deployment Controller is still rolling out pods. However, we don't know whether those
+		// pods have new config or not, in that case we can't determine whether we are progressing or not.
+		// As a result, we fall-back to the previous condition value and hope that in the next sync
+		// the Deployment Controller will be already done rolling out pods.
+		lastCondition := FindOperatorCondition(status.Conditions, conditionName)
+		if lastCondition == nil || lastCondition != nil && lastCondition.Status == operatorv1.ConditionTrue {
+			return true, "Waiting for Deployment to deploy node pods"
+		}
+	}
+
+	return false, ""
+}
+
+func IsDaemonSetProgressing(conditionName string, daemonSet *appsv1.DaemonSet, status *operatorv1.OperatorStatus) (bool, string) {
+	if daemonSet.Generation != daemonSet.Status.ObservedGeneration {
+		return true, "Waiting for DaemonSet to act on changes"
+	}
+
+	// Verify if we there's any node that doesn't contain the daemon pod yet
+	if daemonSet.Status.NumberUnavailable > 0 {
+		// Verify if the current daemonSet generation is the same as the one that this controller
+		// dealt with in the previous sync. If they are not the same, we're definitely rolling
+		// out new config, so we are progressing.
+		lastGeneration := resourcemerge.ExpectedDaemonSetGeneration(daemonSet, status.Generations)
+		if daemonSet.Generation != lastGeneration {
+			return true, fmt.Sprintf("Waiting for %d node pods", daemonSet.Status.NumberUnavailable)
+		}
+
+		// At this point we know two things: we already synced the current generation of the DaemonSet,
+		// but the DaemonSet controller is still rolling out pods. However, we don't know whether those
```

Comment:
same here.  It's not clear to me why the operand is not progressing if the operand has not achieved a steady state.